apps/prairielearn/src/models/lti-user.ts

```diff
@@ -0,0 +1,108 @@
+import { z } from 'zod';
+
+import { callRow } from '@prairielearn/postgres';
+
+import { insertAuditEvent } from './audit-event.js';
+import { ensureEnrollment } from './enrollment.js';
+import { selectInstitutionByShortName } from './institution.js';
+import { insertUserLti, selectUserByUidAndInstitution, updateUserName } from './user.js';
+
+export interface LtiUserResult {
+  userId: string;
+  hasAccess: boolean;
+}
+
+export async function selectOrInsertAndEnrollLtiUser({
+  uid,
+  name,
+  lti_course_instance_id,
+  lti_user_id,
+  lti_context_id,
+  req_date,
+}: {
+  uid: string;
+  name: string | null;
+  lti_course_instance_id: string;
+  lti_user_id: string;
+  lti_context_id: string;
+  req_date: Date;
+}): Promise<LtiUserResult> {
+  const { id: institution_id } = await selectInstitutionByShortName({ short_name: 'LTI' });
+
+  let user = await selectUserByUidAndInstitution({
+    uid,
+    institution_id,
+  });
+
+  if (!user) {
+    user = await insertUserLti({
+      uid,
+      name,
+      lti_course_instance_id,
+      lti_user_id,
+      lti_context_id,
+      institution_id,
+    });
+
+    await insertAuditEvent({
+      action: 'insert',
+      table_name: 'users',
+      row_id: user.user_id,
+      new_row: user,
+      institution_id,
+      course_instance_id: lti_course_instance_id,
+      subject_user_id: user.user_id,
+      agent_authn_user_id: null,
+    });
+  }
+
+  // Update user name if needed
+  if (name !== null && name !== user.name) {
+    const oldUser = user;
+    user = await updateUserName({
+      user_id: user.user_id,
+      name,
+    });
+
+    await insertAuditEvent({
+      action: 'update',
+      action_detail: 'name',
+      table_name: 'users',
+      row_id: user.user_id,
+      old_row: oldUser,
+      new_row: user,
+      institution_id,
+      course_instance_id: lti_course_instance_id,
+      subject_user_id: user.user_id,
+      agent_authn_user_id: null,
+    });
+  }
+
+  const userId = user.user_id;
+
+  if (!userId) {
+    throw new Error('computed NULL user_id');
+  }
```

Comment:
When would this happen?

#12730 is probably relevant here.

Reply:
Yep, this is from a direct translation of the sproc. I didn't want to change semantics in this PR, but it sounds like the preference is to modernize/remove checks that aren't needed when converting a sproc to typescript.